### PR TITLE
fix: standardize token estimation heuristic in streaming handlers

### PR DESCRIPTION
**Pull request by RyanLHicks**
## Summary

`subagent_stream_handler._estimate_tokens` and `event_stream_handler` both used `len(content) // 4` (~4 chars/token), while `BaseAgent.estimate_token_count` uses `math.floor(len(content) / 2.5)` (~2.5 chars/token). The same 100-char input produced 25 vs 40 tokens depending on path, creating inconsistent compaction signals.
This PR aligns both streaming handlers to the 2.5 chars/token heuristic so streaming metrics match what compaction sees.

## Changes
- `subagent_stream_handler._estimate_tokens`: switched to `max(1, math.floor(len(content) / 2.5))`
- `event_stream_handler` tool-args estimate: same formula
- Added `import math` to both files
- New test file `test_token_estimation_consistency.py` with 3 tests covering small/medium/large content and empty input

## Test plan
`pytest tests/agents/test_token_estimation_consistency.py`

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved token estimation accuracy and consistency across the agent system for more reliable usage metrics

* **Tests**
  * Added tests validating token estimation consistency across multiple content sizes and edge cases

### CODE REVIEW

File: code_puppy/tools/file_operations.py [hunk 1]
```diff
@@ -38,7 +43,7 @@ class ListFileOutput(BaseModel):
 
 class ReadFileOutput(BaseModel):
     content: str | None
-    num_tokens: conint(lt=10000)
+    num_tokens: conint(lt=16001)
```
Comment (mpfaffenberger) [inline code_puppy/tools/file_operations.py]: Maybe we should make these configurable

File: code_puppy/tools/file_operations.py [hunk 2]
```diff
@@ -472,6 +477,21 @@ def _read_file(
     if not os.path.isfile(file_path):
         error_msg = f"{file_path} is not a file"
         return ReadFileOutput(content=error_msg, num_tokens=0, error=error_msg)
+
+    # Cache lookup
```
Comment (mpfaffenberger) [inline code_puppy/tools/file_operations.py]: What's the cache lookup for?

File: code_puppy/tools/file_operations.py (hunk 1 above)
Comment (RyanLHicks) [reply code_puppy/tools/file_operations.py]: We can potentially move the token cap and the chars/token ratio into config.py with the current values as defaults. Should they be in their own section or alongside existing model config?

File: code_puppy/tools/file_operations.py (hunk 2 above)
Comment (RyanLHicks) [reply code_puppy/tools/file_operations.py]: It's Primarily session cache to avoid re-reading the same file multiple
times in one session. Keyed by path + line range, invalidated by
mtime so any edit refreshes it.

Framing it as preventative — an internal code_puppy analysis flagged repeated reads as a potential token waste pattern, so the cache is there to short-circuit it if/when it happens. The tests cover cache correctness (hit, mtime invalidation, slice isolation, error bypass) rather than measuring real-world frequency.

Happy to pull it into a separate PR if it's outside the scope of this.

Comment (mpfaffenberger) [reply code_puppy/tools/file_operations.py]: You'll have to explain this further.

It looks like the only thing being guarded by the cache is disk reads. 

On a cache hit, this is still returning the result to the model... so if the model decides to read the same chunk of a file twice, this would have no effect on that situation. 

Furthermore I'm pretty skeptical of multiple reads on the same chunk of a file being a waste of tokens. 

Even if it is, it would be model dependent. This could happen if a compaction occurred and the result of the read was no longer in context. However, the caching mechanism here simply prevents the file from being read from disk.

Let me know if I'm missing something.

File: code_puppy/tools/file_operations.py
Comment (mpfaffenberger) [inline code_puppy/tools/file_operations.py:1]: I think you should remove all changes to this file, and we can move forward with the updates to `subagent_stream_handler` and `event_stream_handler`

File: tests/tools/test_file_operations_coverage.py
Comment (mpfaffenberger) [inline tests/tools/test_file_operations_coverage.py:1]: remove this

File: tests/tools/test_file_operations_extended.py
Comment (mpfaffenberger) [inline tests/tools/test_file_operations_extended.py:1]: remove these changes to adhere to the P/R's title

**Review by mpfaffenberger — CHANGES_REQUESTED** [review]

**Review by mpfaffenberger — APPROVED** [review]

**Comment by mpfaffenberger** [comment]
Now this P/R is looking very good - last thing to do is run linters `uv run ruff format .` and `uv run ruff check --fix`. I need to add instructions to do that in the AGENT.md

**Comment by RyanLHicks** [comment]
Thanks! Pushed the ruff format commit on top.

**Comment by RyanLHicks** [comment]
Two Brief Issues:
1. Existing tests in test_subagent_stream_handler.py were locked to the old len(content)//4 values. Updated to use the new formula.
2. test_token_estimation_consistency.py was instantiating CodePuppyAgent unnecessarily, which couples the test to the full agent init path. Removed the agent and assert against the formula directly.
